design/baremetal-operator/early-cleaning.md

```diff
@@ -0,0 +1,111 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Early out-of-band cleaning
+
+This document proposes running driver-specific out-of-band clean
+steps.
+
+## Status
+
+implementable
+
+## Summary
+
+Depending on the selected driver, BMO will run zero or more out-of-band clean
+steps to ensure the node and the BMC are in the right state. Out-of-band clean
+steps are executed via BMC (e.g. using Redfish protocol) without booting the
+ironic agent (IPA).
+
+The first drivers to receive such handling will be `idrac` and
+`idrac-virtualmedia`. For them we will run the `management.known_good_state`
+clean step that resets iDRAC and clears its job queue.
+
+It is important that these actions run before we try to boot anything on the
+target node, be it the inspection ramdisk or a 3rd party installer ISO.
+
+## Motivation
+
+iDRAC will refuse to create a job if there is already a pending job of the same
+type. To make sure we start in a coherent state, we will reset iDRAC before
+doing anything else to it.
+
+### Goals
+
+* Provide a mechanism to run out-of-band clean steps before inspection.
+* Run `known_good_state` step for iDRAC.
+
+### Non-Goals
+
+* Run any in-band (i.e. using a ramdisk) clean steps.
+* Allow a user to provide the steps to run.
+
+## Proposal
+
+## Design Details
+
+All clean steps run asynchronously in the ironic API, even out-of-band steps
+that are reasonable fast (an order of seconds). When entering the `Registering`
+state, we will check if the driver supplies any reset clean steps. If yes, we
+will start manual cleaning and wait. Otherwise we'll move on.
+
+Resetting won't happen for externally provisioned host.
+
+### Implementation Details/Notes/Constraints
+
+The `AccessDetails` interface will be extended with a new call:
+
+```go
+ResetSteps() []CleanStep
+```
+
+where `CleanStep` is a copy of the corresponding type from gophercloud (to
+avoid adding a dependency on `gophercloud` to `pkg/bmc`).
```

Comment:
If the information in the response depends only on the driver name, and not on the other BMC settings, we could hide all of this inside the ironic provisioner. That should be the only thing that ever asks for the values, and it would just be a `map[string][]nodes.CleanStep` static data structure, right?

Reply:
Aaah, I think I understand now.

Reply:
Although this will require adding driver-dependent code to `pkg/provisioner/ironic`, currently only `pkg/bmc` does it. I'm fine with that, but we need an agreement.

Reply:
Talked on slack a bit, we should probably keep it as it is to avoid unnecessary complicating the provisioner.

Reply:
Yeah, I think it's safe to use a copy of the CleanStep data structure in pkg/bmc. It would be good to describe the other approaches we rejected, and why, as part of this document.